DIPs/DIP1xxx.md

```diff
@@ -0,0 +1,249 @@
+# Expression and Block Statement Template Constraints
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | (number/id -- assigned by DIP Manager)                          |
+| Review Count:   | 0                                                               |
+| Author:         | Nicholas Wilson                                                 |
+| Implementation: | https://github.com/thewilsonator/dmd/tree/template-constraint-dip |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Allow multiple `if` template constraints, for the expression form allow an optional message to be printed in the 
+event that overload resolution fails (similar to `static assert`), as well as a block statement
+form of template constraint that allows the use of `static foreach`.
+That is to say, template constraint `if` becomes the static form of contract precondition `in`.
+
+The template is considered a valid overload iff each of the constraints is satified.
+
+Expression form:
+```D
+template all(alias pred)
+{
+    bool all(Range)(Range range)
+    if (isInputRange!Range)
+    if (is(typeof(unaryFun!pred(range.front))),
+        "`" ~ pred.stringof[1..$-1] ~ "` isn't a unary predicate function for range.front")
+    {
+    }
+}
+```
+
+Block statement form:
+```D
+ptrdiff_t countUntil(alias pred = "a == b", R, Rs...)(R haystack, Rs needles)
+if 
+{
+    static assert(isForwardRange!R);
+    static assert(Rs.length > 0, "need a needle to countUntil with");
```

Comment:
```suggestion
    static assert(Rs.length > 0, "need at least one needle as stop condition");
```
`countUntil` is already part of the full error message (I assume) and we would educate the user that while they are at it they might consider providing several needles instead of calling `countUntil` twice, resulting in better code.